File: tools/generators/generate_e2e_test_vectors.py
#!/usr/bin/env python3
"""
FoxWhisper End-to-End Test Vector Generator
Generates complete protocol flow test vectors for FoxWhisper v0.9
"""

import json
import base64
import os
from typing import Dict, List, Any

class EndToEndTestVectorGenerator:
    """Generates comprehensive end-to-end test vectors"""
    
    def __init__(self):
        self.test_vectors = {}
        
    def generate_handshake_flow(self) -> Dict[str, Any]:
        """Generate complete handshake flow test vectors"""
        
        # Generate cryptographic material
        client_id = base64.b64encode(os.urandom(32)).decode()
        server_id = base64.b64encode(os.urandom(32)).decode()
        
        # X25519 key pairs
        client_x25519_priv = base64.b64encode(os.urandom(32)).decode()
        client_x25519_pub = base64.b64encode(os.urandom(32)).decode()
        server_x25519_priv = base64.b64encode(os.urandom(32)).decode()
        server_x25519_pub = base64.b64encode(os.urandom(32)).decode()
        
        # Kyber material
        client_kyber_pub = base64.b64encode(os.urandom(1568)).decode()
        server_kyber_ciphertext = base64.b64encode(os.urandom(1568)).decode()
        
        # Nonces
        client_nonce = base64.b64encode(os.urandom(16)).decode()
        server_nonce = base64.b64encode(os.urandom(16)).decode()
        
        # Session keys
        session_id = base64.b64encode(os.urandom(32)).decode()
        handshake_hash = base64.b64encode(os.urandom(32)).decode()
        
        handshake_flow = {
            "description": "Complete FoxWhisper handshake flow",
            "participants": ["client", "server"],
            "steps": [
                {
                    "step": 1,
                    "type": "HANDSHAKE_INIT",
                    "from": "client",
                    "to": "server",
                    "message": {
                        "type": "HANDSHAKE_INIT",
                        "version": 1,
                        "client_id": client_id,
                        "x25519_public_key": client_x25519_pub,
                        "kyber_public_key": client_kyber_pub,
                        "timestamp": 1701763200000,
                        "nonce": client_nonce
                    },
                    "expected_response": "HANDSHAKE_RESPONSE"
                },
                {
                    "step": 2,
                    "type": "HANDSHAKE_RESPONSE",
                    "from": "server",
                    "to": "client",
                    "message": {
                        "type": "HANDSHAKE_RESPONSE",
                        "version": 1,
                        "server_id": server_id,
                        "x25519_public_key": server_x25519_pub,
                        "kyber_ciphertext": server_kyber_ciphertext,
                        "timestamp": 1701763201000,
                        "nonce": server_nonce
                    },
                    "expected_response": "HANDSHAKE_COMPLETE"
                },
                {
                    "step": 3,
                    "type": "HANDSHAKE_COMPLETE",
                    "from": "client",
                    "to": "server",
                    "message": {
                        "type": "HANDSHAKE_COMPLETE",
                        "version": 1,
                        "session_id": session_id,
                        "handshake_hash": handshake_hash,
                        "timestamp": 1701763202000
                    },
                    "expected_response": "ENCRYPTED_MESSAGE"
                }
            ],
            "validation_criteria": {
                "all_required_fields_present": True,
                "correct_message_types": True,
                "valid_base64_encoding": True,
                "correct_field_sizes": True,
                "chronological_timestamps": True,
                "matching_session_ids": True
            }
        }
        
        return handshake_flow
    
    def generate_device_addition_flow(self) -> Dict[str, Any]:
        """Generate device addition to existing session"""
        
        existing_device_id = base64.b64encode(os.urandom(32)).decode()
        new_device_id = base64.b64encode(os.urandom(32)).decode()
        session_id = base64.b64encode(os.urandom(32)).decode()
        
        device_addition = {
            "description": "Add new device to existing FoxWhisper session",
            "scenario": "device_addition",
            "existing_session": {
                "session_id": session_id,
                "devices": [existing_device_id]
            },
            "steps": [
                {
                    "step": 1,
                    "type": "DEVICE_ANNOUNCE",
                    "from": "new_device",
                    "to": "existing_devices",
                    "message": {
                        "type": "DEVICE_ANNOUNCE",
                        "version": 1,
                        "device_id": new_device_id,
                        "session_id": session_id,
                        "x25519_public_key": base64.b64encode(os.urandom(32)).decode(),
                        "timestamp": 1701763203000
                    }
                },
                {
                    "step": 2,
                    "type": "DEVICE_ACK",
                    "from": "existing_devices",
                    "to": "new_device",
                    "message": {
                        "type": "DEVICE_ACK",
                        "version": 1,
                        "device_id": existing_device_id,
                        "session_id": session_id,
                        "existing_devices": [existing_device_id],
                        "timestamp": 1701763204000
                    }
                }
            ],
            "validation_criteria": {
                "device_authentication": True,
                "session_consistency": True,
                "no_duplicate_devices": True,
                "proper_device_ids": True
            }
        }
        
        return device_addition
    
    def save_test_vectors(self, filename: str):
        """Save all test vectors to file"""
        
        self.test_vectors["handshake_flow"] = self.generate_handshake_flow()
        self.test_vectors["device_addition"] = self.generate_device_addition_flow()
        
        # Add metadata
        self.test_vectors["_metadata"] = {
            "version": "0.9",
            "generated_by": "FoxWhisper End-to-End Test Vector Generator",
            "description": "Complete protocol flow test vectors for FoxWhisper E2EE",
            "test_categories": ["handshake_flow", "device_addition"],
            "validation_features": [
                "message_structure_validation",
                "field_size_validation",
                "base64_encoding_validation",
                "chronological_validation",
                "session_consistency_validation"
            ]
        }
        
        with open(filename, 'w') as f:
            json.dump(self.test_vectors, f, indent=2)
        
        print(f"✅ End-to-end test vectors saved to {filename}")
        print(f"📊 Generated {len(self.test_vectors)-1} test scenarios")
    
    def validate_test_vectors(self, filename: str):
        """Validate generated test vectors"""
        
        with open(filename, 'r') as f:
            vectors = json.load(f)
        
        validation_results = {
            "handshake_flow": self._validate_handshake_flow(vectors.get("handshake_flow", {})),
            "device_addition": self._validate_device_addition(vectors.get("device_addition", {}))
        }
        
        return validation_results
    
    def _validate_handshake_flow(self, flow: Dict[str, Any]) -> Dict[str, Any]:
        """Validate handshake flow test vector"""
        
        if not flow:
            return {"valid": False, "errors": ["Missing handshake flow"]}
        
        errors = []
        steps = flow.get("steps", [])
        
        # Validate step sequence
        expected_types = ["HANDSHAKE_INIT", "HANDSHAKE_RESPONSE", "HANDSHAKE_COMPLETE"]
        for i, step in enumerate(steps):
            if i >= len(expected_types):
                errors.append(f"Unexpected step {i+1}: {step.get('type')}")
                continue
                
            expected_type = expected_types[i]
            actual_type = step.get("type")
            if actual_type != expected_type:
                errors.append(f"Step {i+1}: expected {expected_type}, got {actual_type}")
        
        # Validate message structure
        for step in steps:
            message = step.get("message", {})
            if not message.get("type"):
                errors.append(f"Step {step.get('step')}: missing message type")
            if not message.get("version"):
                errors.append(f"Step {step.get('step')}: missing version")
            if not message.get("timestamp"):
                errors.append(f"Step {step.get('step')}: missing timestamp")
        
        return {
            "valid": len(errors) == 0,
            "errors": errors,
            "steps_validated": len(steps)
        }
    
    def _validate_device_addition(self, addition: Dict[str, Any]) -> Dict[str, Any]:
        """Validate device addition test vector"""
        
        if not addition:
            return {"valid": False, "errors": ["Missing device addition flow"]}
        
        errors = []
        steps = addition.get("steps", [])
        
        # Validate device announcement
        if len(steps) < 2:
            errors.append("Device addition requires at least 2 steps")
        
        for step in steps:
            message = step.get("message", {})
            if not message.get("device_id"):
                errors.append(f"Step {step.get('step')}: missing device_id")
            if not message.get("session_id"):
                errors.append(f"Step {step.get('step')}: missing session_id")
        
        return {
            "valid": len(errors) == 0,
            "errors": errors,
            "steps_validated": len(steps)
        }

def main():
    print("FoxWhisper End-to-End Test Vector Generator")
    print("=" * 50)
    
    generator = EndToEndTestVectorGenerator()
    
    # Generate test vectors
    output_file = "../test-vectors/handshake/end_to_end_test_vectors.json"
    generator.save_test_vectors(output_file)
    
    # Validate generated vectors
    print("\nValidating generated test vectors...")
    validation_results = generator.validate_test_vectors(output_file)
    
    for category, result in validation_results.items():
        if result["valid"]:
            print(f"✅ {category}: VALID ({result['steps_validated']} steps)")
        else:
            print(f"❌ {category}: INVALID")
            for error in result["errors"]:
                print(f"   Error: {error}")
    
    print(f"\n🎉 End-to-end test vector generation completed!")
    print(f"📁 Saved to: {output_file}")

if __name__ == "__main__":
    main()

File: tools/generators/generate_e2e_test_vectors.rs
use base64::{engine::general_purpose, Engine as _};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::error::Error;
use std::fs;

// FoxWhisper End-to-End Test Vector Generator (Rust)
// Generates complete protocol flow test vectors for FoxWhisper v0.9

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HandshakeMessage {
    #[serde(rename = "type")]
    pub message_type: String,
    pub version: i32,
    #[serde(rename = "client_id", skip_serializing_if = "Option::is_none")]
    pub client_id: Option<String>,
    #[serde(rename = "server_id", skip_serializing_if = "Option::is_none")]
    pub server_id: Option<String>,
    #[serde(rename = "session_id", skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(rename = "x25519_public_key")]
    pub x25519_public_key: String,
    #[serde(rename = "kyber_public_key", skip_serializing_if = "Option::is_none")]
    pub kyber_public_key: Option<String>,
    #[serde(rename = "kyber_ciphertext", skip_serializing_if = "Option::is_none")]
    pub kyber_ciphertext: Option<String>,
    #[serde(rename = "handshake_hash", skip_serializing_if = "Option::is_none")]
    pub handshake_hash: Option<String>,
    pub timestamp: i64,
    #[serde(rename = "nonce", skip_serializing_if = "Option::is_none")]
    pub nonce: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HandshakeStep {
    pub step: i32,
    #[serde(rename = "type")]
    pub step_type: String,
    #[serde(rename = "from")]
    pub from: String,
    #[serde(rename = "to")]
    pub to: String,
    pub message: HandshakeMessage,
    #[serde(rename = "expected_response")]
    pub expected_response: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HandshakeFlow {
    pub description: String,
    pub participants: Vec<String>,
    pub steps: Vec<HandshakeStep>,
}

pub struct EndToEndTestVectorGenerator {
    test_vectors: HashMap<String, serde_json::Value>,
}

impl EndToEndTestVectorGenerator {
    pub fn new() -> Self {
        Self {
            test_vectors: HashMap::new(),
        }
    }

    pub fn generate_handshake_flow(&mut self) -> HandshakeFlow {
        // Generate cryptographic material
        let client_id = Some(generate_random_base64(32));
        let server_id = Some(generate_random_base64(32));
        let client_x25519_pub = generate_random_base64(32);
        let server_x25519_pub = generate_random_base64(32);
        let client_kyber_pub = Some(generate_random_base64(1568));
        let server_kyber_cipher = Some(generate_random_base64(1568));
        let client_nonce = Some(generate_random_base64(16));
        let server_nonce = Some(generate_random_base64(16));
        let session_id = Some(generate_random_base64(32));
        let handshake_hash = Some(generate_random_base64(32));

        HandshakeFlow {
            description: "Complete FoxWhisper handshake flow".to_string(),
            participants: vec!["client".to_string(), "server".to_string()],
            steps: vec![
                HandshakeStep {
                    step: 1,
                    step_type: "HANDSHAKE_INIT".to_string(),
                    from: "client".to_string(),
                    to: "server".to_string(),
                    message: HandshakeMessage {
                        message_type: "HANDSHAKE_INIT".to_string(),
                        version: 1,
                        client_id: client_id.clone(),
                        server_id: None,
                        session_id: None,
                        x25519_public_key: client_x25519_pub,
                        kyber_public_key: client_kyber_pub,
                        kyber_ciphertext: None,
                        handshake_hash: None,
                        timestamp: 1701763200000,
                        nonce: client_nonce,
                    },
                    expected_response: "HANDSHAKE_RESPONSE".to_string(),
                },
                HandshakeStep {
                    step: 2,
                    step_type: "HANDSHAKE_RESPONSE".to_string(),
                    from: "server".to_string(),
                    to: "client".to_string(),
                    message: HandshakeMessage {
                        message_type: "HANDSHAKE_RESPONSE".to_string(),
                        version: 1,
                        client_id: None,
                        server_id: server_id.clone(),
                        session_id: None,
                        x25519_public_key: server_x25519_pub,
                        kyber_public_key: None,
                        kyber_ciphertext: server_kyber_cipher,
                        handshake_hash: None,
                        timestamp: 1701763201000,
                        nonce: server_nonce,
                    },
                    expected_response: "HANDSHAKE_COMPLETE".to_string(),
                },
                HandshakeStep {
                    step: 3,
                    step_type: "HANDSHAKE_COMPLETE".to_string(),
                    from: "client".to_string(),
                    to: "server".to_string(),
                    message: HandshakeMessage {
                        message_type: "HANDSHAKE_COMPLETE".to_string(),
                        version: 1,
                        client_id: None,
                        server_id: None,
                        session_id: session_id,
                        x25519_public_key: String::new(), // Empty for this message type
                        kyber_public_key: None,
                        kyber_ciphertext: None,
                        handshake_hash: handshake_hash,
                        timestamp: 1701763202000,
                        nonce: None,
                    },
                    expected_response: "ENCRYPTED_MESSAGE".to_string(),
                },
            ],
        }
    }

    pub fn save_test_vectors(&mut self, filename: &str) -> Result<(), Box<dyn Error>> {
        let handshake_flow = self.generate_handshake_flow();
        self.test_vectors.insert(
            "handshake_flow".to_string(),
            serde_json::to_value(&handshake_flow)?,
        );

        // Add metadata
        let metadata = serde_json::json!({
            "version": "0.9",
            "generated_by": "FoxWhisper End-to-End Test Vector Generator (Rust)",
            "description": "Complete protocol flow test vectors for FoxWhisper E2EE",
            "test_categories": ["handshake_flow"],
            "validation_features": [
                "message_structure_validation",
                "field_size_validation",
                "base64_encoding_validation",
                "chronological_validation",
                "session_consistency_validation"
            ]
        });

        self.test_vectors.insert("_metadata".to_string(), metadata);

        // Save to file
        let json_data = serde_json::to_string_pretty(&self.test_vectors)?;
        fs::write(filename, json_data)?;

        println!("✅ End-to-end test vectors saved to {}", filename);
        println!(
            "📊 Generated {} test scenarios",
            self.test_vectors.len() - 1
        );

        Ok(())
    }
}

fn generate_random_base64(size: usize) -> String {
    use rand::thread_rng;
    use rand::RngCore;

    let mut bytes = vec![0u8; size];
    let mut rng = thread_rng();
    rng.fill_bytes(&mut bytes);
    general_purpose::STANDARD.encode(bytes)
}

fn main() -> Result<(), Box<dyn Error>> {
    println!("FoxWhisper End-to-End Test Vector Generator (Rust)");
    println!("{}", "=".repeat(50));

    let mut generator = EndToEndTestVectorGenerator::new();

    // Generate test vectors
    let output_file = "../test-vectors/handshake/end_to_end_test_vectors_rust.json";
    generator.save_test_vectors(output_file)?;

    println!("\n🎉 End-to-end test vector generation completed!");
    println!("📁 Saved to: {}", output_file);

    Ok(())
}
